### PR TITLE
Switch to Scala 2.12.7

### PR DESCRIPTION
**Pull request by chetanmeh**
Updates the Scala compiler and library to 2.12.7

## Description

Apart from switching the jar versions to 2.12 some code changes were required to avoid use of now deprecated features. 

* [Avoid ambiguty between global executionContext and local one](https://stackoverflow.com/questions/40442717/implicit-executioncontext-priority-in-scala-2-12)
* [Use explicit eta expansion instead of implicit one][2]
* Use JavaConverters over JavaConversion
* Remove use of [deprecated future methods][3]

## Related issue and scope

- [x] I opened an issue to propose and discuss this change (#3952)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

[2]: https://stackoverflow.com/questions/45657747/deprecation-warning-when-compiling-eta-expansion-of-zero-argument-method
[3]: https://viktorklang.com/blog/Futures-in-Scala-2.12-part-5.html

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/database/memory/MemoryArtifactStore.scala [hunk 1]
```diff
@@ -131,14 +131,13 @@ class MemoryArtifactStore[DocumentAbstraction <: DocumentSerializer](dbName: Str
...
       case _: DocumentConflictException =>
         transid.finished(this, start, s"[PUT] '$dbName', document: '$docinfoStr'; conflict.")
-    })
+      case _ =>
+    }
 
-    f.onSuccess({
-      case _ => transid.finished(this, start, s"[PUT] '$dbName' completed document: '$docinfoStr'")
-    })
+    f.foreach(_ => transid.finished(this, start, s"[PUT] '$dbName' completed document: '$docinfoStr'"))
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/memory/MemoryArtifactStore.scala]: Can we use `onComplete` everywhere there's the success/failure case?

**Comment by chetanmeh** [comment]
Compilation time for whole project via reduced from 2 min 39 sec to 1 min 46 sec, thus a 33% improvement !!

```
$./gradlew clean
$./gradlew compileScala compileTestScala 
```

File: common/scala/src/main/scala/whisk/core/database/memory/MemoryArtifactStore.scala (hunk 1 above)
Comment (chetanmeh) [reply common/scala/src/main/scala/whisk/core/database/memory/MemoryArtifactStore.scala]: Done

**Comment by chetanmeh** [comment]
Compilation time is reduced specially for test module

| 2.11 | 2.12 |
| ---- | ----- |
| image|image|

**Comment by chetanmeh** [comment]
Created 16 PR :cold_sweat: and no  code change needed in other repos. Of 16 build is failing for 4 for reasons not related to this change. 

Once this PR is merged those 16 PR would need to be updated to point to latest master and then merged

**Comment by chetanmeh** [comment]
@vvraskin @cbickel Can you check if your downstream project can compile fine with this change. If required you can use the PR branch

```
git clone --depth=1 --single-branch -b scala-2-12 https://github.com/chetanmeh/incubator-openwhisk.git openwhisk
```

**Comment by markusthoemmes** [comment]
Kudos @chetanmeh this is a lot of repetitive work. Thanks a ton for pulling through!

**Comment by csantanapr** [comment]
Thanks @chetanmeh for this

**Comment by chetanmeh** [comment]
@csantanapr I have reverted all git branch change from various PR. So all PR now only have required changes for Scala 2.12 support.

**Comment by csantanapr** [comment]
In IBM we did some refactoring in our internal repos and public repos for this and we are ready for the Scala bump.

**Comment by csantanapr** [comment]
The conflict needs to be resolved before we can start the merge fest.

**Comment by chetanmeh** [comment]
Post rebase the invoker is throwing exception at runtime

```
Exception in thread "pool-1-thread-3" java.lang.AbstractMethodError: Method whisk/core/connector/CompletionMessage.whisk$core$connector$Message$_setter_$transid_$eq(Lwhisk/common/TransactionMetadata;)V is abstract
	at whisk.core.connector.CompletionMessage.whisk$core$connector$Message$_setter_$transid_$eq(Message.scala)
	at whisk.core.connector.Message.$init$(Message.scala:31)
	at whisk.core.connector.AcknowledegmentMessage.<init>(Message.scala:78)
	at whisk.core.connector.CompletionMessage.<init>(Message.scala:93)
	at whisk.core.invoker.InvokerReactive.whisk$core$invoker$InvokerReactive$$send$1(InvokerReactive.scala:128)
	at whisk.core.invoker.InvokerReactive.$anonfun$ack$1(InvokerReactive.scala:149)
	at whisk.core.invoker.InvokerReactive.$anonfun$ack$1$adapted(InvokerReactive.scala:121)
	at whisk.core.containerpool.ContainerProxy.$anonfun$initializeAndRun$15(ContainerProxy.scala:428)
	at scala.util.Success.foreach(Try.scala:249)
	at scala.concurrent.Future.$anonfun$foreach$1$adapted(Future.scala:225)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Strangely compilation passes fine. The `transid` in `Message` is getting marked as abstract even though its provided in case class parameters

Seems like this was changed in #4041 and may be a bug in Scala compiler. Needs further investigation

**Comment by chetanmeh** [comment]
The issue can be seen with [this fiddle][1]. It passes with 2.11 but fails with 2.12. Probably related to scala/bug#10477

[1]: https://scastie.scala-lang.org/0KHtPAZTTqSHqBANVy52Bg

File: common/scala/src/main/scala/whisk/core/connector/Message.scala [hunk 2]
```diff
@@ -75,8 +75,8 @@ object ActivationMessage extends DefaultJsonProtocol {
  * Message that is sent from the invoker to the controller after action is completed or after slot is free again for
  * new actions.
  */
-abstract class AcknowledegmentMessage() extends Message {
-  override val transid: TransactionId
+abstract class AcknowledegmentMessage(private val tid: TransactionId) extends Message {
```
Comment (chetanmeh) [inline common/scala/src/main/scala/whisk/core/connector/Message.scala]: Had to change code here to workaround Scala compiler bug (see previous comments)

**Comment by chetanmeh** [comment]
@csantanapr  The PR is green now. So we can now do the merge

**Comment by csantanapr** [comment]
thanks @chetanmeh 👍

**Comment by chetanmeh** [comment]
Thanks @csantanapr for taking care of merge and changes in all other repos

**Pull request merged**